### ISSUE TITLE
compute_energy_ctm_2site silently drops bonds whose RDM contracts to zero (both can die -> E = 0.0 exactly, and the iPEPS smoke test passes on it)

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#845]
> 🤖 **AI-generated re-scope** — written by Claude Code, posted by @yingjerkao.

## ⚠️ Re-scoped 2026-08-11 — this is much wider than originally filed

The guard added in #847 made the failure audible, and it immediately fired in the repo's **own default test path**. Two claims in the original report below are wrong, both in the direction that made this look like a corner case.

**Wrong claim 1 — "degenerate χ=D² corner".** χ=D² is not the gate. Scanning D ∈ {2,3} × χ ∈ {4,5,6,9,12} × SU steps ∈ {3,20} × CTM `max_iter` ∈ {3,50}, on a **real** state with the plain Heisenberg gate:

| D | χ | χ=D²? | steps | E | dead bonds |
|---|---|---|---|---|---|
| 2 | 4 | yes | 3 | **0.000000000000** | `_rdm1x2_2site`, `_rdm2x1_2site` |
| 2 | 4 | yes | 20 | −0.253220115327 | `_rdm1x2_2site` |
| 2 | 5 | no | 20 | −0.515247274080 | — |
| 3 | 4 | **no** | 20 | +0.108774182541 | `_rdm2x1_2site` |
| 3 | 5 | **no** | 3 | **0.000000000000** | `_rdm1x2_2site`, `_rdm2x1_2site` |
| 3 | 6 | no | 20 | +0.221471397915 | — |
| 3 | 9 | yes | 3 | −0.003178078804 | `_rdm2x1_2site` |
| 3 | 12 | no | 20 | +0.324763761667 | — |

It fires at D=3 χ=4 and D=3 χ=5 — neither is D² — and does **not** fire at D=3 χ=6 or χ=12. Raising `max_iter` from 3 to 50 changes nothing in any dead-bond row, so this is structural, not a CTM convergence shortfall.

**Wrong claim 2 — needs a complex state.** Every row above is real. The original text hedged that real "reproduces given the right state"; in fact real is the common case, and the complex χ=4 run I filed against was just the first one I happened to see.

## The part that matters most

**Both bonds can die at once, and then `E` is exactly `0.0`.**

```
tests/test_ipeps.py::TestIPEPSRun::test_ipeps_runs_without_error
    iPEPSConfig(max_bond_dim=2, num_imaginary_steps=3, dt=0.1,
                ctm=CTMConfig(chi=4, max_iter=3))
    -> energy repr : 0.0
    -> exactly 0.0 : True
    -> assert jnp.isfinite(energy)  ->  True   # passes
```

That is this repository's end-to-end iPEPS smoke test. It returns an energy of exactly zero for a Heisenberg model and passes, because `isfinite(0.0)` is true. `::test_lbfgs_optimizer_runs` and `::test_cg_optimizer_runs` trigger the same warning on both `_rdm2x1_2site` and `_rdm1x2_2site`.

So the tests that exist to prove iPEPS works end-to-end have been proving it against a fully collapsed energy.

## What this changes

- **#847 is not a fix, and was never claimed to be one** — it warns, it does not repair. That was the right call, but it was made on a severity estimate that was too low. The underlying defect needs an actual fix.
- **The smoke tests need a real assertion.** `isfinite` cannot distinguish a converged Heisenberg energy from `0.0`. A negative-energy floor, or a trace check on the RDMs, would have caught this years of commits ago.
- **Positive energies appear throughout the scan** (+0.109, +0.221, +0.325) at short SU depth. Some of that is legitimately unconverged simple update; the dead-bond rows are the separable, reliable signal.

## Still true from the original report

The mechanism, and the fact that nothing upstream can see it: the CTM convergence criterion compares corner *singular values*, which are identical between the two bases, so it reports `converged=True` with `diff ~ 1e-17` while a bond's RDM network contracts to zero.

---

<details>
<summary>Original report, as filed (retained for the measurement record)</summary>

> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

`compute_energy_ctm_2site` sums `E_h + E_v` with no validity check on either reduced density matrix. On a degenerate corner one of the two RDM networks can contract to **exactly zero**, and that bond then contributes `0` to the energy instead of its real value. The returned number is wrong by nearly a factor of two, and every status signal in the stack reports success.

Found while fixing #842; unrelated to that fix (it reproduces on the real path too, given the right state — real vs complex is only the perturbation that selects which fixed point you land on).

## Measurement

D=2 Heisenberg simple-update state at tau=10 (200 steps x dt=0.05), `sublattice_rotate_gate(heisenberg_gate())`. The *same* state is used for every row; the "complex" column is that state cast to `complex128`, which is physically identical.

Project venv, jax 0.10.2 / numpy 2.4.4, state `sha256[:16]=3faa697b2eb7c7d2`, reproducible across 5 fresh processes.

| chi | corner spectrum `s0/s_last` | E (real) | E (complex) | \|dE\| | both report |
|---|---|---|---|---|---|
| **4** | **1.00** (exactly flat) | -0.548635274240 | **-0.274984329862** | **2.7e-01** | `converged=True`, diff ~ 1e-17, n_iter=2 |
| 5 | 1.01e+04 | -0.548643264430 | -0.548661710416 | 1.8e-05 | `converged=False`, 200 iters |
| 6 | 8.17e+05 | -0.548661443911 | -0.548644519609 | 1.7e-05 | `converged=False`, 200 iters |
| 8 | 8.29e+05 | -0.548586755277 | -0.548815318564 | 2.3e-04 | `converged=False`, 200 iters |

chi=5, 6, 8 and chi=4-real all cluster at -0.5486. The chi=4 complex run is low by a factor of 1.995.

## Mechanism

Decomposed per bond direction:

```
chi=4 real: E_h=-0.273650944378  E_v=-0.274984329862   tr(rdm_h)=1.000000  tr(rdm_v)=1.000000
chi=4 cplx: E_h=+0.000000000000  E_v=-0.274984329862   tr(rdm_h)=0.000000  tr(rdm_v)=1.000000
chi=6 cplx: E_h=-0.273656654340  E_v=-0.274987865268   tr(rdm_h)=1.000000  tr(rdm_v)=1.000000
```

The horizontal RDM is not merely ill-conditioned, it is identically zero:

```
real: ||rdm_h||_F=9.986948e-01  tr=+1.000000e+00   ||rdm_v||_F=9.988269e-01  tr=+1.000000e+00
cplx: ||rdm_h||_F=0.000000e+00  tr=+0.000000e+00   ||rdm_v||_F=9.988269e-01  tr=+1.000000e+00
```

`rdm_v` is bit-identical between the two runs. Only the horizontal bond dies, and it contributes exactly `0.0` to the sum. The ratio 1.995 is just `(E_h + E_v) / E_v`.

## Why nothing catches it

Three independent layers each decline to notice:

1. **The corner is exactly degenerate.** At D=2, chi=4 (`chi = D^2`, the untruncated case) the corner spectrum is `[0.25, 0.25, 0.25, 0.25]`, so the eigenbasis is arbitrary. Real and complex LAPACK select different ones -- the corner diagonals are `[0, -1, -1, 0]` vs `[0, 0, -1, -1]`, a permutation -- and the environments differ by **O(1)** (`B.C1 = 1.00e+00`).

2. **The convergence criterion is blind by construction here.** `_ctm_sv_diff` compares corner *singular values*, which are `[0.25]x4` in **both** bases -- identical. It reports `diff = 2.78e-17, converged=True` after 2 sweeps for an environment that is unusable. This is not a tolerance that needs tightening; the criterion cannot see this failure at any tolerance. Related: #780/#781 (gauge-dependent criteria), and #839, which added the `converged` flag -- note this is a case where `converged=True` and the environment is *still* invalid, so that flag alone was never going to catch it.

3. **The energy path has no trace check.** `ipeps_rdm.py:313-315` is `E_h = einsum(...); E_v = einsum(...); return (E_h + E_v).real`. There is no assertion that `tr(rdm) ~ 1`. `_normalise_rdm` (`_ctm_tensor_energy.py:31`) already contemplates this case -- its docstring says the fallback applies "when the trace has effectively vanished relative to the matrix, where there is no physical content left to preserve" -- and its zero-matrix guard is what turns `0/0` into a clean `0.0` rather than NaN. That guard is correct in itself, but the caller then adds "no physical content left" into the energy as though it were a bond contribution of zero.

Note also `check_ctm_env` (`_ctm_diagnostics.py:88`) exists but has **zero callers** in `src/`.

## Suggested fix

The cheap, robust guard is (3), and it is independent of whether (1)/(2) are ever addressed: check `|tr(rdm) - 1| < tol` for each RDM before contracting it with `H`, and warn or raise rather than silently contributing `0`. A zero-trace RDM is unambiguously not a density matrix -- there is no state for which that is the right answer.

Warning (not raising) matches the #747/#839 policy for a non-converged CTM, and would have made this loud instead of a plausible-looking number 2x off.

## Established / not established

**Established:** at `chi = D^2 = 4` the corner spectrum is exactly flat; a perturbation as small as real-vs-complex arithmetic flips which fixed point the sweep lands on; one RDM then contracts to exactly zero; and the convergence flag, the RDM normaliser and the energy sum all report success.

**Not established:** that this reaches a *truncating* configuration (`chi > D^2`). At chi >= 5 the corner spectrum is well-conditioned (ratio 1e4-1e6) and the two paths agree to ~1e-5, consistent with the known unconverged floor -- note none of those runs converge at all (diff 1.5e-2 to 4.3e-2 after 200 sweeps), which is a separate open question about the criterion.

**Excluded from the above:** two early probe runs executed with a cwd where `uv run` resolves `~/miniforge3` (jax 0.10.0 / numpy 2.2.3) rather than the project venv. Those produced a different simple-update state and made the discrepancy disappear. All numbers reported here are the project venv only.

## Repro

```python
import warnings, jax, jax.numpy as jnp
jax.config.update("jax_enable_x64", True)
from tenax.algorithms.ipeps_config import CTMConfig, iPEPSConfig
from tenax.algorithms.ipeps_ctm_convergence import ctm_2site
from tenax.algorithms.ipeps_rdm import _rdm2x1_2site, _rdm1x2_2site
from tenax.algorithms.ipeps import heisenberg_gate, ipeps, sublattice_rotate_gate

gate_t = sublattice_rotate_gate(heisenberg_gate()); gate = gate_t.todense()
with warnings.catch_warnings():
    warnings.simplefilter("ignore")
    _E, (A, B), _ = ipeps(gate_t, None, iPEPSConfig(
        max_bond_dim=2, num_imaginary_steps=200, dt=0.05, unit_cell="1x1",
        ctm=CTMConfig(chi=6, max_iter=100, conv_tol=1e-10)))

Ar, Br = A.todense(), B.todense()
Ac, Bc = Ar.astype(jnp.complex128), Br.astype(jnp.complex128)
cfg = CTMConfig(chi=4, chi_I=4, max_iter=200, conv_tol=1e-10)

for tag, (X, Y) in (("real", (Ar, Br)), ("cplx", (Ac, Bc))):
    eA, eB = ctm_2site(X, Y, cfg)
    rh = _rdm2x1_2site(X, Y, eA, eB, 2)
    print(tag, "||rdm_h||_F =", float(jnp.linalg.norm(rh)),
          " tr =", complex(jnp.einsum("ijij->", rh)).real)
# real 0.9986948...  tr = 1.0
# cplx 0.0           tr = 0.0
```

Requires #842's fix (on `main` via #843) for the complex branch to run at all.

</details>

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Re-scoped the title and body above. Short version of what changed:

- **χ=D² is not the trigger.** It fires at D=3 χ=4 and D=3 χ=5, and does not fire at D=3 χ=6 or χ=12. Raising CTM `max_iter` 3 → 50 changes nothing, so it is structural rather than a convergence shortfall.
- **A complex state is not required.** The whole scan is real; the complex χ=4 run this was filed against was just the first instance I saw.
- **Both bonds can die at once, giving `E` exactly `0.0`** — including in `tests/test_ipeps.py::TestIPEPSRun::test_ipeps_runs_without_error`, which passes because it asserts only `jnp.isfinite(energy)`, and `isfinite(0.0)` is true. `test_lbfgs_optimizer_runs` and `test_cg_optimizer_runs` warn on both bonds too.

The guard from #847 reports this; it does not repair it. Keeping this open for the actual fix, and the smoke-test assertions need strengthening separately — `isfinite` cannot tell a converged Heisenberg energy from zero.
